### PR TITLE
Fixes #13711 - Do not validate subnet domain relation

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

File: app/models/subnet_domain.rb [hunk 1]
```diff
@@ -2,9 +2,6 @@ class SubnetDomain < ActiveRecord::Base
   belongs_to :domain
   belongs_to :subnet
 
-  validates :subnet_id, :presence => true
-  validates :domain_id, :presence => true, :uniqueness => {:scope => :subnet_id}
```
Comment (ohadlevy) [inline app/models/subnet_domain.rb:6]: how do you validate it then? IMHO this needs to stay and error should propagate to the calling object if its not already?

Comment (ares) [inline app/models/subnet_domain.rb:6]: This validates internal data that are not provided by user. Active record already catches invalid ids so we don't need to validate anything. What we really need is DB constraint, I'll add the migration to make both fields required, I was not aware it's missing as well. Makes sense?

Comment (ohadlevy) [inline app/models/subnet_domain.rb:6]: I'm not sure, where would an api user will get an error if the domain /
subnet combination is not unique? I dont want to get a FK error in an api
call.

Comment (ares) [inline app/models/subnet_domain.rb:6]: How would you create that duplication through subnet API? Either the subnet exist, then you only change its attribute, if you specify the same domain id, AR will handle that. Or it doesn't exists and you create new subnet (hence no duplication either). As stated in issue description, `SubnetDomain` object is never used directly and AFAIK there's no API to create these associations.

**Comment by ares** [comment]
[test]

**Comment by ares** [comment]
[test] foreman tasks were fixed

**Comment by dLobatog** [comment]
@ares Can you explain a bit better what's the issue? What I understood from the Redmine issue:
- You try to create a Subnet with an attached Domain on the web UI
- Click on submit - Domain is unselected but no errors are shown. 
- You see on the logs SubnetDomain is missing some foreign ID (presumably Subnet's?), can you confirm which foreign ID is missing? 

Perhaps you can validate the relation object is present (`:subnet` and `:domain`). not the ID and see if the error goes away. Do you get the error without removing the validations and just including the unique db index?

I think having the validation in both places is what's considered the best practice here, validation on the model to avoid a potential network query that will error out and have a nice error message, and validation on the database to guarantee data consistency.

**Comment by ares** [comment]
The issue is not in displaying of error, that's just top of it. Try to create a **new** subnet with some domain associated and see yourself that it does not work. It does not work because the validation in SubnetDomain is wrong. We could fix this by adding a condition that would skip the validation on new records, but the original cause is that the validation should not be in this model in the first place. This model is just relation entity that is never used directly, it's something that ActiveRecord uses in `has_many ... :through =>` association. Validations should validate user data only. Therefore I'm removing it which not only fixes the issue but improves the code.

EDIT: just to clarify it more, if there's a place where validation should exist, it would be a Subnet model. Validation would check that `domain_ids` is an array and that is contains only numbers and these numbers are valid ids but all of that is already handled by ActiveRecord and obviously we don't do such validations on other models...

**Comment by ares** [comment]
> You see on the logs SubnetDomain is missing some foreign ID (presumably Subnet's?), can you confirm which foreign ID is missing?

yes, it's subnets id being nil since it's not persisted yet

>  Do you get the error without removing the validations and just including the unique db index?

no, the validation prevents saving...

**Comment by iNecas** [comment]
Just to make it clear: this validation it wrong, because the ids are not available by the time of the creating the subnet.

I agree with Marek, that in  a join-like model, as the SubnetDomain is, it makes a little sense to have active record validations: this are not the data the user supplies, unless there is a situation where user data can directly cause the integrity failure without some other validation catching it. The integrity itself should be properly be handled by the foreign keys. If there is a code error that invalidates the integrity, I don't think the nice validation error is needed (the user would not know what to do anyway).

When using `habtm` (I don't really recommend doing it:), there are no validations there either.

**Comment by dLobatog** [comment]
@ares The validations here are wrong on #new because they use _ids, which is not there when it's being created. I don't argue that was wrong. What I suggest is defining the validations and relations properly. Validations should validate the object, and has_many through should use inverse_of:

On subnet.rb

``` ruby
has_many :subnet_domains, :dependent => :destroy, :inverse_of => :subnet
```

On domain.rb

``` ruby
has_many :subnet_domains, :dependent => :destroy, :inverse_of => :domain
```

On subnet_domain.rb

``` ruby
  validates :subnet, :presence => true
  validates :domain, :presence => true, :uniqueness => {:scope => :subnet}
```

That will allow you to create new subnets with domains. We have a similar structure on user_role and taxable_taxonomy. Again, having the database validations helps too in case of race conditions and the index will make it faster.

Here's another argument for having both validations I mentioned before:

Validating this in the model saves you a database query (possibly across the network) that will error out, and doing it in the database guarantees data consistency. Remember ActiveRecord doesn't honor validations in all contexts (update_attribute, update_all, decrement, etc...) - especially when join models are involved.

If what bugs you is the duplication of code, I found this gem [schema_validations](https://github.com/SchemaPlus/schema_validations) that takes validations from the db and injects them in activerecord too.

cc @iNecas - precisely because we're doing an uniqueness check here, we cannot use habtm and we want to validate stuff.

**Comment by ares** [comment]
> That will allow you to create new subnets with domains. We have a similar structure on user_role and taxable_taxonomy. Again, having the database validations helps too in case of race conditions and the index will make it faster.

Validations for uniqueness won't help with race conditions (rails don't handle race conditions in uniq validations, they just do one query, if it return no record, they run second to create it).

> Validating this in the model saves you a database query

Uniqueness validation actually triggers extra DB query to verify that no conflicting record exists.

> If what bugs you is the duplication of code

What bugs me is that validations should not be used in this case. They should validate user input, not something our application (AR) generates. Please show me an example of how user can trigger this validation failure. Note, that even if this validation fails, user don't get any useful information, since it's registered on relation, not the failing object itself.

**Comment by ares** [comment]
I don't want to remove WoC myself but this is no longer waiting for contributor I think.

**Comment by domcleal** [comment]
I agree that the validations are of limited use as the DB index is going to be more reliable and reduce the number of queries.  That said, the validation errors can show up to the user in a friendlier way than the DB constraint errors, though probably not for subnet/domain associations in develop.

> What bugs me is that validations should not be used in this case. They should validate user input, not something our application (AR) generates. Please show me an example of how user can trigger this validation failure.

Probably the clearest place you might see it is when doing something like this:

<pre>
2.0.0-p353 :026 > subnet.attributes = {:domains => [Domain.find(12)]*2}
2016-03-02T13:42:34 [sql] [D]   Domain Load (0.1ms)  SELECT  "domains".* FROM "domains"  WHERE "domains"."id" = ?  ORDER BY domains.name LIMIT 1  [["id", 12]]
2016-03-02T13:42:34 [sql] [D]   Domain Load (0.1ms)  SELECT "domains".* FROM "domains" INNER JOIN "subnet_domains" ON "domains"."id" = "subnet_domains"."domain_id" WHERE "subnet_domains"."subnet_id" = ?  ORDER BY domains.name  [["subnet_id", 1]]
2016-03-02T13:42:34 [sql] [D]    (0.1ms)  begin transaction
2016-03-02T13:42:34 [sql] [D]   SQL (0.2ms)  DELETE FROM "subnet_domains" WHERE "subnet_domains"."subnet_id" = ? AND "subnet_domains"."domain_id" = 1  [["subnet_id", 1]]
2016-03-02T13:42:34 [sql] [D]   SubnetDomain Exists (0.1ms)  SELECT  1 AS one FROM "subnet_domains"  WHERE ("subnet_domains"."domain_id" = 12 AND "subnet_domains"."subnet_id" = 1) LIMIT 1
2016-03-02T13:42:34 [sql] [D]   SQL (0.1ms)  INSERT INTO "subnet_domains" ("created_at", "domain_id", "subnet_id", "updated_at") VALUES (?, ?, ?, ?)  [["created_at", "2016-03-02 13:42:34.803385"], ["domain_id", 12], ["subnet_id", 1], ["updated_at", "2016-03-02 13:42:34.803385"]]
2016-03-02T13:42:34 [sql] [D]   SubnetDomain Exists (0.1ms)  SELECT  1 AS one FROM "subnet_domains"  WHERE ("subnet_domains"."domain_id" = 12 AND "subnet_domains"."subnet_id" = 1) LIMIT 1
2016-03-02T13:42:34 [sql] [D]    (0.1ms)  rollback transaction
ActiveRecord::RecordInvalid: Validation failed: »Domain« has already been taken
</pre>

Relying on just on the index will raise an error from the database (e.g. `SQLite3::ConstraintException: UNIQUE constraint failed: subnet_domains.subnet_id, subnet_domains.domain_id: ...`).

I don't think setting `subnet.domains` is possible via the API since 3326499 because it's protected (only `subnet.domain_ids` is permitted, which doesn't allow this), but on many other models with these types of joins, these attributes _can_ be set.  Plugins may still use this (such as foreman_setup), even if protected at the moment from users.

So just for consistency across models I think we should keep the validators in the style that @dLobatog suggests in https://github.com/theforeman/foreman/pull/3186#issuecomment-185180092.

**Comment by ares** [comment]
Ok, I guess I was outvoted :-) here you are.

**Comment by domcleal** [comment]
Thanks @ares.  Would you mind adding a unit test for the subnet model to check it can be created with domains?

**Comment by ares** [comment]
Adding only because I really want this bug fixed. I really doubt this test should be part of our test suite, unless we want to be CI for ActiveRecord. Btw I also removed one empty test.

**Comment by domcleal** [comment]
Merged as 46eec8b, thanks @ares.

**Pull request closed without merging**